### DISCUSSION TITLE
Nelder-Mead optimizer

### DISCUSSION THREAD

**Discussion by kathmcp** [phoebe-project/phoebe2#508] [Questions]
Hi Kyle,

I am now trying to run the Nelder-Mead optimizer, but it seems to be running very, very slowly. It is taking about 20s per iteration, and then after 18 iterations it seems to have just stopped. I removed the orb and mesh datasets, so it's only running one light curve, but this didn't seem to help. I also tried using fast compute with ellc, but it doesn't work for contact binaries. I am trying to just vary the mass ratio, initially. 

When I first enter the parameters for my binary, I get the following errors and warnings (although it does run_compute later)

Wed, 28 Jul 2021 16:30 BUNDLE       ERROR   Constraint 'pot@contact_envelope@envelope@constraint' raised the following error while attempting to solve for 'pot@contact_envelope@envelope@component'.  Consider flipping the constraint or changing the value of one of ['requiv@primary@star@component', 'q@binary@orbit@component', 'sma@binary@orbit@component', 'pot@contact_envelope@envelope@component'] until the constraint succeeds.  Original error: roche_contact_Omega_at_partial_vol::Volume is outside bounds.
Wed, 28 Jul 2021 16:30 BUNDLE       ERROR   Constraint 'requiv@secondary@star@constraint' raised the following error while attempting to solve for 'requiv@secondary@star@component'.  Consider flipping the constraint or changing the value of one of ['pot@contact_envelope@envelope@component', 'q@binary@orbit@component', 'sma@binary@orbit@component', 'requiv@secondary@star@component'] until the constraint succeeds.  Original error: potential > L1 critical value
Wed, 28 Jul 2021 16:30 BUNDLE       ERROR   Constraint 'fillout_factor@contact_envelope@envelope@constraint' raised the following error while setting the value of 'fillout_factor@contact_envelope@envelope@component'.  Original error: value of fillout_factor=-13.410026693369053 not within limits of [<Quantity 0.>, <Quantity 1.>]
Wed, 28 Jul 2021 16:30 BUNDLE       WARNING constraints ['pot@contact_envelope@envelope@constraint'] failed to run.  Address errors and try again.  Call run_failed_constraints to see the tracebacks.  If not addressed, this warning will continue to be raised and will throw an error at run_compute.
Wed, 28 Jul 2021 16:30 BUNDLE       WARNING primary is underflowing at L1 and not a contact system (requiv=1.5, requiv_min=1.8346231503528594, requiv_max=1.9639267680718653)  If not addressed, this warning will continue to be raised and will throw an error at run_compute.
Wed, 28 Jul 2021 16:30 BUNDLE       WARNING secondary is overflowing at L2/L3 (requiv=1.4999999999999996, requiv_min=1.2725418568681297, requiv_max=0.9342870720587503)  If not addressed, this warning will continue to be raised and will throw an error at run_compute.

Are these errors and warnings contributing to the problems with the optimizer?

**Comment by kecnry** [comment]
These again are probably intermediate warnings.  As long as the system passes [b.run_checks](http://phoebe-project.org/docs/latest/api/phoebe.frontend.bundle.Bundle.run_checks) and the original forward model runs successfully, it shouldn't be an issue.

How long does run_compute take?  Contact binaries can be quite expensive, and each iteration in Nelder-Mead will require several calls to run_compute, so I'm not surprised if it takes quite a while.  It's probably well worth the effort to try to optimize your forward model as much as possible first.  [This tutorial from the latest workshop](http://phoebe-project.org/workshops/2021june/Tutorial_optimizing_computations) covers some of the things you might want to try.

**Reply by kathmcp** [reply]
So, b.run_checks() give status = PASS, and the first run_compute takes about 9 seconds to run, so I assume that means all the errors and warnings after entering the system parameters don't matter. Then I try adding in the optimiser and when it gets to run_solver it goes really slowly and eventually seems to stop. I will have a look at the tutorial you mentioned and see if that helps. Thanks again! :)

**Reply by kecnry** [reply]
Ok, great!  I think that should help you speed things up a bit.  By seems to stop do you mean it completes before you expected or that the progress bar gets stuck and it stalls entirely?

**Reply by kathmcp** [reply]
The progress bar gets stuck at the 18th iteration and it seems to stall entirely. I have left it for maybe 10minutes or so and it is still at iteration 18

**Reply by kecnry** [reply]
Ok... then I don't think speeding it up will likely fix that (although still will be useful to do once it is working).  My guess is that the optimizer is wandering into an area of the parameter space where building the mesh is hanging.  

Can you enable a debugging logger and send me the output?  Immediately before run_solver, add `phoebe.logger(flevel='DEBUG', filename='debug.log')` (and remove any previous logger statements).  Then just run the script, wait for it to stall again, and send me the last 50 or so lines of that log file.

If you can also send me a copy of the bundle immediately before calling run_solver, that would be easiest for me to try to reproduce.  If you don't want to attach it publicly here, you can send it to me by email.

Thanks!

**Reply by kathmcp** [reply]
Hi Kyle, here are the files, hope it works this time! Safe travels!
![snip1](https://user-images.githubusercontent.com/81473857/127396440-d4701595-66b1-49bd-b476-5ec7b420ce03.PNG)
![snip2](https://user-images.githubusercontent.com/81473857/127396446-7dc32f6e-d8eb-40c2-8037-c87c937086ca.PNG)
![snip3](https://user-images.githubusercontent.com/81473857/127396447-73e45a0c-9c27-4564-affb-af510f144394.PNG)
![snip4](https://user-images.githubusercontent.com/81473857/127396450-c992027d-575f-46f8-9f85-0aac0cb59506.PNG)
![snip5](https://user-images.githubusercontent.com/81473857/127396451-f52e109b-2d20-4519-aad3-5640f6bb32a5.PNG)
![snip6](https://user-images.githubusercontent.com/81473857/127396452-fd78549c-5e46-4650-a563-e4c9eb3bd322.PNG)
![snip7](https://user-images.githubusercontent.com/81473857/127396454-3be8d1d9-54f8-4d33-b201-e74752bcd659.PNG)

**Comment by kathmcp** [comment]
Hi Kyle,

I have now added in the code for the debugging logger, and although it made a file, it didn't write anything to it. Please find attached snapshots of my file so far. I have tried to keep it simple by commenting out some of the parameters I entered that I don't think are so important to see if that helped. And so far I only have one of my 3 passbands in the file. I hope the snapshots are enough for you to be able to see what I am trying to do. If you need anything else, please let me know.

Thanks again,
Kathryn

________________________________

**Reply by kecnry** [reply]
Neither of the files attached correctly (that seems to happen if attaching as a reply by email to the discussion).  Can you either try attaching them from GitHub directly and/or just email them to me.  I'm traveling over the next few days, but will try to look into it as soon as I can.

**Comment by kathmcp** [comment]
Here is the csv file with the observations in it, if that's of any help.
Thanks,
Kathryn

________________________________

**Comment by kathmcp** [comment]
Hi Kyle, sorry, I thought I had attached the files in here last night, but they are not here, so here they are again, along with an xls file of the observations if you need it. thanks!
[BFilterAllFluxTimeFluxErr2 - Copy.xlsx](https://github.com/phoebe-project/phoebe2/files/6900501/BFilterAllFluxTimeFluxErr2.-.Copy.xlsx)

![snip1](https://user-images.githubusercontent.com/81473857/127489735-93096b89-c46a-434a-a378-cb3750ca20ae.PNG)
![snip2](https://user-images.githubusercontent.com/81473857/127489738-3681d751-1174-4b5b-8bcc-4902a75efd3d.PNG)
![snip3](https://user-images.githubusercontent.com/81473857/127489740-1207eb9d-7483-4091-a293-427299165397.PNG)
![snip4](https://user-images.githubusercontent.com/81473857/127489743-c91b4759-78b5-4697-bd3c-f0593fb17639.PNG)
![snip5](https://user-images.githubusercontent.com/81473857/127489744-46d246e0-a4b9-4fa9-8ad0-bec58d65c3ac.PNG)
![snip6](https://user-images.githubusercontent.com/81473857/127489748-b6726fd6-6da0-4964-86ec-de1649b81cea.PNG)
![snip7](https://user-images.githubusercontent.com/81473857/127489750-3977bce2-ef2f-47e7-8b1d-1de7462a060b.PNG)

**Reply by kathmcp** [reply]
Hi Kyle, I have added in a merit function and managed to get it run up to 47% so far...
![snip8](https://user-images.githubusercontent.com/81473857/127643086-6c87a996-68b8-4bcc-b22e-cd1c9dddbc2f.PNG)
![snip9](https://user-images.githubusercontent.com/81473857/127643089-730a2b8d-ff9e-4b91-8c83-b4af2cf8007e.PNG)

**Reply by kecnry** [reply]
ok, then that reinforces the idea that its just getting stuck in an area of parameter space that your priors now avoid (at least a little longer).  Its hard to reproduce from screenshots though... can you please call `b_cb.save('filename.bundle')` immediately before calling `b_cb.run_solver` and send that file?

**Reply by kathmcp** [reply]
Hopefully this works, I had to zip it to attach it here.
[CB_LD_pblum_NM.zip](https://github.com/phoebe-project/phoebe2/files/6907798/CB_LD_pblum_NM.zip)

**Reply by kecnry** [reply]
Great, I was able to run the solver and it finished successfully on my machine after 48 iterations (so unfortunately unable to reproduce the issue you're seeing). 

What version of phoebe are you using (`print(phoebe.__version__)`)?

The log may not have written to the file since it stalled.  Can you try instead using a log printed to the screen and copy/paste the last 10 or so lines?  Just call `phoebe.logger('DEBUG')` immediately before calling `run_solver`.

**Reply by kathmcp** [reply]
The PHOEBE version is 2.3.49. And here are what seems to be the last few lines of the log...It ran for about 25minutes doing this and then seems to have stopped here

Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   ld_func=logarithmic, ld_coeffs=[0.846 0.095], atm=ck2004, ldatm=none
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   secondary._populate_lc(dataset=lc01, ignore_effects=False)
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   ld_func=logarithmic, ld_coeffs=[0.846 0.095], atm=ck2004, ldatm=none
Fri, 30 Jul 2021 15:37 BACKENDS     DEBUG   rank:0/1 PhoebeBackend._run_single_time: filling packets at time=2459309.442
Fri, 30 Jul 2021 15:37 BACKENDS     DEBUG   rank:0/1 PhoebeBackend._run_single_time: returning packetlist at time=2459309.442
Fri, 30 Jul 2021 15:37 BACKENDS     DEBUG   rank:0/1 PhoebeBackend._run_single_time(i=60, time=2459309.447, infolist=[{'dataset': 'lc01', 'component': None, 'kind': 'lc', 'needs_mesh': True}], **kwargs.keys=dict_keys(['system', 'hier', 'meshablerefs', 'starrefs', 'dynamics_method', 'ts', 'xs', 'ys', 'zs', 'vxs', 'vys', 'vzs', 'ethetas', 'elongans', 'eincls']))
Fri, 30 Jul 2021 15:37 BACKENDS     DEBUG   rank:0/1 PhoebeBackend._run_single_time: extracting dynamics at time=2459309.447
Fri, 30 Jul 2021 15:37 BACKENDS     DEBUG   rank:0/1 PhoebeBackend._run_single_time: calling system.update_positions at time=2459309.447
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   system.update_positions ignore_effects=False
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   primary.update_position ignore_effects=False
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   primary.update_position: accessing standard mesh at t=2459309.447
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   primary.update_position: placing in orbit, Mesh.from_scaledproto at t=2459309.447
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   primary.update_position: calling compute_local_quantities at t=2459309.447 ignore_effects=False
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   primary._fill_loggs
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   primary._fill_gravs
Fri, 30 Jul 2021 15:37 UNIVERSE     DEBUG   primary.instantaneous_gpole

**Reply by kecnry** [reply]
That is a bizarre spot for it to stall, is it possible that your machine is running out of memory?  Can you open a system monitor and see if your memory is filling up?

If that isn't the obvious culprit, then I'll ask some other members of the team and see if we can get a machine to reproduce the issue.  Unfortunately, its difficult for me to track down and fix until I can reproduce it (without iterating back and forth and adding more debug statements in the code for you to run).  Sorry about the troubles!

**Reply by kathmcp** [reply]
Yes it looks like memory might have been the problem that time, although I don't think it was before, but maybe. I am freeing up some memory now, so will try again.

However, in the meantime I updated phoebe to version 2.3.50, and now run_compute is not working. Should I uninstall phoebe and reinstall to get back to 2.3.49?

**Reply by kecnry** [reply]
Ok, I'll try to profile the memory usage and see if there are any leaks (but that will take a while and I'm not sure when we'll get the chance to do that).  As for the upgrade, nothing in 2.3.50 should have affected run_compute.  I'd suggest just re-installing in case something went wrong, and if its giving an error message, post that as a new bug report please.

**Reply by kathmcp** [reply]
Hi Kyle, I have uninstalled and reinstalled phoebe, but I can't seem to get it to go back to v2.3.49, so am persevering with v2.3.50. When I run the file with the parameters I sent you before (q=0.15, sma=1.85, requiv@primary=1.05), it fails 2 checks at run_checks, and the warnings tell me requiv needs to be between 1.27 and 1.67. I then change requiv to 1.3, and have tried various combinations of sma and q, but I keep getting the following errors and warnings. Do I just need to keep adjusting these parameters until something works?

WARNING constraints ['pot@contact_envelope@envelope@constraint'] failed to run.  Address errors and try again.  Call run_failed_constraints to see the tracebacks.  If not addressed, this warning will continue to be raised and will throw an error at run_compute.
ERROR   Constraint 'pot@contact_envelope@envelope@constraint' raised the following error while attempting to solve for 'pot@contact_envelope@envelope@component'.  Consider flipping the constraint or changing the value of one of ['requiv@primary@star@component', 'q@binary@orbit@component', 'sma@binary@orbit@component', 'pot@contact_envelope@envelope@component'] until the constraint succeeds.  Original error: cannot convert object with physical_type=length to solar units
Sat, 31 Jul 2021 11:54 BUNDLE       ERROR   Constraint 'requiv@secondary@star@constraint' raised the following error while attempting to solve for 'requiv@secondary@star@component'.  Consider flipping the constraint or changing the value of one of ['pot@contact_envelope@envelope@component', 'q@binary@orbit@component', 'sma@binary@orbit@component', 'requiv@secondary@star@component'] until the constraint succeeds.  Original error: cannot convert object with physical_type=dimensionless to solar units

**Reply by kathmcp** [reply]
Hi again Kyle, I am thinking that if it ran on your computer, but doesn't run on mine, I might try buying a new laptop (we would need a new one soon anyway). What specifications would be needed for PHOEBE? It will be another Windows system, as I am not familiar with Mac and don't have time to learn. I would be grateful for any advice :)

**Reply by kathmcp** [reply]
Hi Kyle, I have bought a new computer now, and reinstalled everything (WSL2, Ubuntu and Miniconda3, and then PHOEBE) following the instructions for Conda Environments from here http://phoebe-project.org/install/latest/linux/auto. Everything seems to work ok, but the file is still not running. I have tried version 2.3.50 and 2.3.49, but neither works. As far as I can see I haven't changed any inputs, so I have no idea why run_compute() ran before and isn't running now. I am still getting the errors above, and also this one below now. When I check what value of asini it has calculated after my initial input parameters, it still has 3.5 (from default values I think). Then later after some more inputs (and after run_compute(), it says asini failed to run (the second error noted below). There are numerous other errors too. If I can send you anything to help, please feel free to ask. Thanks again for any help!

ERROR   Constraint 'asini@binary@orbit@constraint' raised the following error while attempting to solve for 'asini@binary@orbit@component'.  Consider flipping the constraint or changing the value of one of ['sma@binary@orbit@component', 'incl@binary@orbit@component', 'asini@binary@orbit@component'] until the constraint succeeds.  Original error: cannot convert object with physical_type=length to solar units

WARNING constraints ['asini@binary@orbit@constraint'] failed to run.  Address errors and try again.  Call run_failed_constraints to see the tracebacks.  If not addressed, this warning will continue to be raised and will throw an error at run_compute.

**Reply by kecnry** [reply]
The issue you're seeing now is still related to the bugfix you posted (#509) - the underlying cause seems to have been an update to astropy which broke the way phoebe handled parsing and converting units in constraints.  I am testing a fix for that now and will try to release an updated 2.3.51 which should get that working for you.  See that issue or keep a lookout for the release for more details.

Hopefully the new machine with more memory will help with the stalling issues you had once that is out!

**Reply by kathmcp** [reply]
Ok fantastic, thank you!
